### PR TITLE
GSoC2020 Enhancing the 2D Arrangement Demo

### PR DESCRIPTION
**Pull request by theartful**
## Summary of Changes

This PR mainly improves the support for Algebraic arrangements, and adds the support for Bezier arrangements and rational function arrangements. There is also a number of clean-ups and minor features. It is a continuation on the work done last year (GSoC 2019) by Saurabh Singh. 

## Release Management

* Affected package(s): 2D Arrangements
* License and copyright ownership: The license used by CGAL

### CODE REVIEW

**Comment by lrineau** [comment]
This PR was tested two nights ago, and the results are mixed:
https://cgal.geometryfactory.com/CGAL/testsuite/results-5.2-Ic-6.shtml#Arrangement_on_surface_2_Demo

There is also that error that might be due to this PR:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-6/Surface_sweep_2/TestReport_Blake_Windows_MSVC-2019-Community-Release-64bits.gz

**Comment by theartful** [comment]
Working on addressing these problems.

**Comment by theartful** [comment]
I managed to locally reproduce and fix all the compilation errors in the log except for https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-6/Arrangement_on_surface_2_Demo/TestReport_lrineau_Fedora-with-LEDA.gz
The arrangement package also fails for the same platform with the same errors
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-6/Arrangement_on_surface_2/TestReport_lrineau_Fedora-with-LEDA.gz
Please try testing again.

File: Arrangement_on_surface_2/demo/Arrangement_on_surface_2/AlgebraicCurveParser.cpp [hunk 1]
```diff
@@ -146,7 +146,7 @@ AlgebraicCurveParser<Polynomial_d>::operator()(const std::string& expression)
     return {};
 }
 
-#ifdef CGAL_USE_CORE
+#ifdef CGAL_USE_Core
```
Comment (lrineau) [inline Arrangement_on_surface_2/demo/Arrangement_on_surface_2/AlgebraicCurveParser.cpp:149]: I disagree. In CGAL, there are 189 occurrences of `CGAL_USE_CORE` and 0 of `CGAL_USE_Core`.

**Review by lrineau — CHANGES_REQUESTED** [review]
Please change back `CGAL_USE_Core` to `CGAL_USE_CORE`.

**Comment by theartful** [comment]
@lrineau  I just tried installing CGAL with CGAL_WITH_Core=ON flag to test it, and the compiler was passed "CGAL_USE_Core" not "CGAL_USE_CORE", so as far as I can tell "CGAL_USE_CORE" is incorrect.
I also tried "CGAL_WITH_Core=OFF" and compile the demo, and sure enough the compilation was faulty. Making this change fixed the issue.

**Comment by theartful** [comment]
Actually, nevermind I will check this again.

**Comment by theartful** [comment]
Error on my end, you're right.

**Review by lrineau — APPROVED** [review]

**Comment by maxGimeno** [comment]
errors : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-15/Arrangement_on_surface_2_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2015-Debug-64bits.gz

**Comment by theartful** [comment]
msvc "WinDef.h" defines a macro for "min" and "max" which causes problems. I surrounded all the uses of min/max with parenthesis in the demo to solve the problem, but now the same issue occurred with code in arrangement package:
https://github.com/CGAL/cgal/blob/afc0abb82f341e5dadaae545dc3ac51d30a8f5dc/Arrangement_on_surface_2/include/CGAL/Curved_kernel_via_analysis_2/gfx/Curve_renderer_traits.h#L58
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-16/Arrangement_on_surface_2_Demo/TestReport_Blake_Windows_MSVC-2019-Community-Release-64bits.gz
Should I also do this in the package or "#define NOMINMAX" which should prevent defining these macros?

**Comment by maxGimeno** [comment]
> I surrounded all the uses of min/max with parenthesis

That's how we do too.

> Should I also do this in the package (change the variable name / surround with parenthesis when appropriate) or "#define NOMINMAX" which should prevent defining these macros?

Use the parenthesis, please

**Comment by efifogel** [comment]
Just rename the local var.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
> Should I also do this in the package (change the variable name / surround with parenthesis when appropriate) or "#define NOMINMAX" which should prevent defining these macros?

A library cannot `#define NOMINMAX`. Wherever you put that definition, user code might already have included a header before that includes `WinDef.h`. That is why we have to use the workarounds. Efi is right, in your case, renaming the variable is probably the easiest workaround.

**Comment by maxGimeno** [comment]
There are some errors on Windows and a timeout on ubuntu : https://cgal.geometryfactory.com/CGAL/testsuite/results-5.2-Ic-17.shtml#Arrangement_on_surface_2_Demo

**Comment by theartful** [comment]
I will ignore the "internal compiler error" thing, and will try to make the compilation faster and try to reproduce the min/max stuff locally to better address it.

**Comment by maxGimeno** [comment]
thanks.

**Comment by theartful** [comment]
My last commits should address the compile time issue. Let's try testing this again.

**Comment by maxGimeno** [comment]
errors: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-20/Arrangement_on_surface_2_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2015-Debug-64bits.gz

**Comment by theartful** [comment]
Let's try this one more time. I have it working locally with msvc2015 and msvc2017 without warnings.

**Comment by maxGimeno** [comment]
Warnings  : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-25/Arrangement_on_surface_2_Demo/TestReport_lrineau_Fedora-with-LEDA.gz
and error: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-25/Arrangement_on_surface_2_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2017-Release-64bits.gz

**Comment by theartful** [comment]
The error message does not help "internal compiler error". I looked it up, and it's due to optimization failure on msvc. As far as I can tell this is not related to the demo since it also happens with the arrangements examples tests (different platform though). I tried reproducing it locally but failed.
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-25/Arrangement_on_surface_2_Examples/TestReport_lrineau_Ubuntu-GCC_master_CXX20-Release.gz

The warnings are also not related to the demo, but I can look into them.

**Comment by maxGimeno** [comment]
I think it is related, as it is the only Arrangement related PR that was tested last night.

**Comment by theartful** [comment]
I highly doubt that these changes are the cause; the changes are in the algebraic traits and without them there are warnings for almost all platforms + "gfx" package responsible for rendering algebraic curves is only exposed in this version of the demo and not in the examples so these warnings didn't show up before. The warnings in question are in Bezier and Rational traits.
I'm suspecting this is because there is some code exposed by the demo that is not by the examples.
Under any case, you're right, the PR is definitely responsible for the warnings and I will try to solve them. I will ignore the MSVC error however, since I couldn't reproduce it locally using msvc2017 and the flags used locally.

**Comment by theartful** [comment]
@lrineau You are right.
https://github.com/CGAL/cgal/blob/ae8496fdfca4d3814d2b507d398e373848d32fb8/Arrangement_on_surface_2/include/CGAL/Curved_kernel_via_analysis_2/Arc_2.h#L2974-L2976
This change is responsible for Arrangement examples failing (I thought they failed before as well).

gcc fixed the bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=97347
I reproduced the error locally using the commit right before the fix https://pastebin.com/ksVjBpzu and couldn't reproduce it using the current gcc master head, so I expect the error to disappear when gcc is updated in the test suite.

Regarding the warnings, I couldn't reproduce them using any of the versions of gcc that caused them in the test suite, however, I believe my last commit should fix them.

Sorry this PR is taking this many rounds, but I hope we can finish it as I will be unavailable very soon.

**Comment by maxGimeno** [comment]
>Sorry this PR is taking this many rounds

Don't worry, it happends quite often, and you're doing a great job trying to fix things :+1:

**Comment by lrineau** [comment]
> Sorry this PR is taking this many rounds, but I hope we can finish it as I will be unavailable soon.

As far as I can see, you did a great job so far. Thanks a lot.

**Comment by efifogel** [comment]
Achmed, I second Maime and Laurent.
Your efforts are much appreciated!
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
@theartful I see at https://cgal.geometryfactory.com/CGAL/Members/wiki/Google_Summer_of_Code#Accepted_projects_proposal that we have received the electronic version of the contributor license agreement. Have you also send the paper version of it?

**Comment by theartful** [comment]
I have not. I will send it as soon as I can.

**Comment by maxGimeno** [comment]
The warnings are still there but we decided to ignore them for now. Let's integrate this PR:

Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.2-Ic-27.shtml

**Comment by lrineau** [comment]
Thanks @theartful for your contribution during that GSoC'2020. If you are still willing to fix the warnings, you can open a new PR with the same branch (if the branch has new commits).

**Pull request merged**